Implement Observable#forEach

maraisr opened the pull request:
#### 7b5785e176c2c2c256da316e0ad378e927a404b9
<pre>
Implement Observable#forEach
<a href="https://bugs.webkit.org/show_bug.cgi?id=277510">https://bugs.webkit.org/show_bug.cgi?id=277510</a>

Reviewed by Chris Dumez.

This change introduces the `.forEach` promise-returning operator
for Observables, using the InternalObserver and subscribes immediately.
This is required by the specification <a href="https://wicg.github.io/observable/#dom-observable-foreach.">https://wicg.github.io/observable/#dom-observable-foreach.</a>

This change also includes a new VisitorCallback IDL which represents a callback
that receives each value of an Subscription, which unlike MapperCallback, does not return a value.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-forEach.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-forEach.any.worker-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-forEach.window-expected.txt: Rebaselined.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverForEach.cpp: Added.
(WebCore::createInternalObserverOperatorForEach):
* Source/WebCore/dom/InternalObserverForEach.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::forEach):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
  Exposes the `Promise&lt;undefiend&gt; forEach(callback, options)`.
* Source/WebCore/dom/VisitorCallback.h: Added.
* Source/WebCore/dom/VisitorCallback.idl: Added.

Canonical link: <a href="https://commits.webkit.org/285693@main">https://commits.webkit.org/285693@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/693aa8ed9ee5a4b26fb4daee94abd7d5750bf21a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52949 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26328 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77795 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24758 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62080 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/734 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57771 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16201 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76587 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47868 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63260 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38182 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44483 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20746 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23089 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66270 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21095 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79406 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/836 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/322 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66166 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/977 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63272 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65445 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16176 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9306 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7482 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/799 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/829 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/815 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/835 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

maraisr commented on Source/WebCore/dom/InternalObserverForEach.cpp:
```diff
@@ -0,0 +1,143 @@
+        promise->reject<IDLAny>(reason);
+    });
+
+    auto observer = InternalObserverForEach::create(context, callback, dependantSignal, promise);
```
Note for reviewer: Although all our tests pass, I do not feel this line is correct.

We really should have:

```diff
- auto observer = InternalObserverForEach::create(context, callback, dependantSignal, promise);
+ auto observer = InternalObserverForEach::create(context, callback, signal, promise);
```

as in, pass the `signal` as opposed to `dependantSignal` to mirror [the spec](https://wicg.github.io/observable/#dom-observable-foreach) with these _points of interest_:

_"Let visitor callback controller be a new AbortController."_

and then further

_"If an exception E was thrown, then reject p with E, and signal abort visitor callback controller with E."_

I interpreted that as to `abort` on the `signal` we created here, than abort on the `dependantSignal`.

**BUT!**

I was unable to get the tests to pass... You'll find the tests _"visitor callback rejection microtask ordering"_ and _"Aborting the passed-in signal rejects the returned promise"_ to fail, if all we do is swap `dependantSignal` here with `signal` — and 🤷🏻 no idea why.

A simple test to async abort the signal, that it rejects the promise — fails.

```js
let o = new Observable(() => {});
let controller = new AbortController()
let prom = o.forEach(() => {}, { signal: controller.signal };

controller.abort('error');

// NOTE: We hang here, neither reject _nor_ resolve.
try {
  await prom;
  unreachable();
} catch (e) {
  assert(e, 'error');
}
```

Through many trial and errors, I found that:

```cpp
dependantSignal->addAlgorithm([promise](JSC::JSValue reason) {
  promise->reject<IDLAny>(reason);
});
```

would never-ever fire.

But would if we passed `dependantSignal` down into `InternalObserverForEach`. Which to me feel's wrong.

So, does `AbortSignal::any` do something with how the wiring works, so that if a dependent signal aborts, the "result signal" does not run its algorithms? Or maybe this just C ownership things?

maraisr replied:
I did some further investigation, and I am almost fully convinced that either the spec is _wrong_ — or wpt tests are wrong. The desired behaviour we want (where handlers all fire as wpt wants), is not specified anymore (see https://github.com/whatwg/dom/pull/1152).

As for the above comment, the reason the tests was failing with the dependent signal approach, was largely due to dependent signals not forwarding one signals abort status to all other sibling signals. 

You can try this yourself with:

```js
const a = new AbortController();
const b = new AbortController();
const c = AbortSignal.any([a.signal, b.signal]);

a.signal.addEventListener('abort', () => console.log('a aborted')); // DOES NOT CALL
b.signal.addEventListener('abort', () => console.log('b aborted')); // called
c.addEventListener('abort', () => console.log('c aborted')); // called

b.abort();

console.log('a', a.signal.aborted); // false <~ aborting b does not abort a
console.log('b', b.signal.aborted); // true
console.log('c', c.aborted); // true
```

Which in our case would be that `a` is our internal abort signal, `b` the `SubscribeOptions.signal`, and `c` the [deprived signal point 3](https://wicg.github.io/observable/#dom-observable-foreach) tells us to create.

So for wpt tests that check that passed-in abort signals ran if we aborted, is not how dependent signals work.

As for why the algorithm wouldn't run (from my comment above), when sending `signal` as opposed to `dependantSignal` is still a mystery. But at least with the latest changes, the code reads a lot more sensibly, and passes the same tests — though isn't reflective of the spec.

We want all promise returning operators to have a predictable abort and promise timing order — so  I will work with @domenic, @domfarolino and @keithamus on making those spec changes to reflect this, and wpt test consistency.

Given this feature is still behind a feature flag, and the wpt tests pass on both Chrome and _now us_ — I am inclined to not let that spec change halt landing this PR.

maraisr replied:
^ with all that being said, I also know very little C++ (kind of learning as im going) and might be entirely likely that the ordering is a symptom of AbortSignals being passed as references, vs pointers. Or something.

domfarolino replied:
I understand this response is entirely late, and I apologize for this getting lost in my inbox, but I'd like to follow-up here to see if there is anything I can do, and to make sure I understand your original concern. In the first comment in this thread, you believed that the internal observer should be created with `signal` instead of `dependentSignal`, but you couldn't do this without failing some tests. But I noticed the PR landed with you passing in `signal`, and the tests in question passing: https://github.com/WebKit/WebKit/pull/32099/files#diff-f5ec8493ab9017c0d501d5725e74b79bbe13541fad819b65ee4569bbdaa05418R6.

Unfortunately there is no commit history in this PR for me to detect what you might have changed in between writing this thread and merging the PR to get it to work, so I'm curious — what did you change to fix it?

The only thing that the signal passed into the internal observer is used for is to get aborted in the case ["next" error case](https://wicg.github.io/observable/#ref-for-abortcontroller-signal-abort%E2%91%A1). The spec uses AbortController to achieve this since we shouldn't abort AbortSignals directly (although it seems that's fine in WebKit: https://bugs.webkit.org/show_bug.cgi?id=282069). So it shouldn't actually matter which signal you pass into the internal observer: `signal` or `dependentSignal`. In fact, it seems like `signal` doesn't even need to exist? I feel like you can just create `dependentSignal` directly based on `options.signal` — [`signal` itself](https://github.com/WebKit/WebKit/pull/32099/files#diff-3b52f4377c8ea8a505ebfc45f0d8bd7e6e5388a6eb762fa7dca8ac12405936e9R121) seems like it's just an empty pass-through, which isn't adding any direct value. Is that right? In that case, I wonder if you can get rid of it and just pass `dependentSignal` directly into the internal observer's `::create()` method.

maraisr commented on Source/WebCore/dom/InternalObserverForEach.cpp:
```diff
@@ -0,0 +1,136 @@
+
+void createInternalObserverOperatorForEach(ScriptExecutionContext& context, Ref<Observable> observable, Ref<VisitorCallback> callback, SubscribeOptions options, Ref<DeferredPromise>&& promise)
+{
+    Vector<Ref<AbortSignal>> dependentSignals = { };
```
as per https://github.com/WebKit/WebKit/pull/35261#discussion_r1804687496

maraisr commented on Source/WebCore/dom/InternalObserverForEach.cpp:
```diff
@@ -0,0 +1,137 @@
+
+void createInternalObserverOperatorForEach(ScriptExecutionContext& context, Observable& observable, Ref<VisitorCallback>&& callback, const SubscribeOptions& options, Ref<DeferredPromise>&& promise)
+{
+    Ref signal = AbortSignal::create(&context);
```
As per https://bugs.webkit.org/show_bug.cgi?id=282069, save's coming back to this later.

cdumez commented on Source/WebCore/dom/InternalObserverForEach.cpp:
```diff
@@ -0,0 +1,137 @@
+
+    void error(JSC::JSValue value) final
+    {
+        Ref { m_promise }->reject<IDLAny>(value);
```
Our rules of thumb in WebKit is that if you have to call `Ref { m_promise }` more than once, then your should be adding a protected getter:
```
Ref<DeferredPromise> protectedPromise() const { return m_promise; }
```
then call `protectedPromise()->reject()` here. Same comment applies elsewhere.

maraisr replied:
Given we let this go through in https://github.com/WebKit/WebKit/pull/35261/files#diff-42a899eeb66d8a4163f94ff20bdc4d0cfeaf4fb2c7a3e1a3ba69906c29f9e741R65 are you okay for me to action this as part of https://bugs.webkit.org/show_bug.cgi?id=282074 for a broader cleanup of the Observable code?

cdumez replied:
I'd rather we correct this PR before landing.

maraisr replied:
Sounds good to me, Ill do this right away!

maraisr replied:
I've pushed that change, please let me know if I've put those protected getter's in the right place (in terms of order). I want to learn the ordering convention.

cdumez commented on Source/WebCore/dom/InternalObserverForEach.cpp:
```diff
@@ -0,0 +1,137 @@
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
+    {
+        Ref { m_callback }->visitJSFunction(visitor);
```
`Ref { m_callback }` is called repeatedly too.

cdumez reviewed (approved):
LGTM with nits.

cdumez reviewed (approved):
Yep, this looks good.

maraisr committed:
Implement Observable#forEach
https://bugs.webkit.org/show_bug.cgi?id=277510

Reviewed by Chris Dumez.

This change introduces the `.forEach` promise-returning operator
for Observables, using the InternalObserver and subscribes immediately.
This is required by the specification https://wicg.github.io/observable/#dom-observable-foreach.

This change also includes a new VisitorCallback IDL which represents a callback
that receives each value of an Subscription, which unlike MapperCallback, does not return a value.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-forEach.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-forEach.any.worker-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-forEach.window-expected.txt: Rebaselined.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverForEach.cpp: Added.
(WebCore::createInternalObserverOperatorForEach):
* Source/WebCore/dom/InternalObserverForEach.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::forEach):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
  Exposes the `Promise<undefiend> forEach(callback, options)`.
* Source/WebCore/dom/VisitorCallback.h: Added.
* Source/WebCore/dom/VisitorCallback.idl: Added.

Canonical link: https://commits.webkit.org/285693@main
